zap/src/irgen/mod.rs

```diff
@@ -272,6 +274,109 @@ pub trait Gen {
 			}
 		}
 	}
+
+	fn new_scope(&mut self);
+	fn current_scope(&mut self) -> &mut Scope;
+	fn end_scope(&mut self);
+
+	fn get_bitpack(&mut self) -> (BitpackMask, Var) {
+		let scope = self.current_scope();
+
+		let existing = scope
+			.bitpack_budget
+			.last_mut()
+			.filter(|(shift, _)| *shift < (BitpackMask::BITS as u8 - 1));
+		if let Some(existing) = existing {
+			existing.0 += 1;
+			(1 << existing.0, Var::Name(existing.1.clone()))
+		} else {
+			let (name, _) = self.add_occurrence("bool");
+			self.current_scope().bitpack_budget.push((0, name.clone()));
+			(1, Var::Name(name))
+		}
+	}
+
+	fn variant_storage(&mut self, amount: usize) -> VariantStorageKind {
+		// 0 is variant 1, 1 is variant 2
+		if amount <= 2 {
+			VariantStorageKind::Bit(self.get_bitpack())
+		} else if self.current_scope().remaining_bitpack_budget() as usize >= amount {
+			VariantStorageKind::Bitpack(std::iter::repeat_with(|| self.get_bitpack()).take(amount).collect())
+		} else {
+			VariantStorageKind::Full(NumTy::from_f64(0.0, amount as f64 - 1.0))
+		}
+	}
+}
+
+#[derive(Debug)]
+pub enum OutputEntryKind {
+	Stmt(Stmt),
+	Buffer(OutputBuffer),
+}
+
+#[derive(Debug)]
+pub struct OutputEntry(OutputEntryKind);
```

Comment:
Why did you wrap `OutputEntryKind` in a `struct`? It shouldn't need to be wrapped.